cruft/_commands/utils/generate.py

```diff
@@ -0,0 +1,120 @@
+
+from cookiecutter.generate import generate_files
+from git import Repo
+
+from .cookiecutter import CookiecutterContext, generate_cookiecutter_context
+from .cruft import get_cruft_file
+
+try:
+    import toml  # type: ignore
+except ImportError:  # pragma: no cover
+    toml = None  # type: ignore
+
+CruftState = Dict[str, Any]
+
+
+def cookiecutter_template(
+    output_dir: Path,
+    repo: Repo,
+    project_dir: Path = Path("."),
+    cookiecutter_input: bool = False,
+    checkout: Optional[str] = None,
+) -> CookiecutterContext:
+    """Generate a clean cookiecutter template in output_dir."""
+    pyproject_file = project_dir / "pyproject.toml"
+    cruft_file = get_cruft_file(project_dir)
+    cruft_state = json.loads(cruft_file.read_text())
+    commit = checkout or repo.remotes.origin.refs["HEAD"]
+
+    repo.head.reset(commit=commit, working_tree=True)
+
+    output_dir.mkdir(parents=True, exist_ok=True)
+    context = _generate_output(cruft_state, Path(repo.working_dir), cookiecutter_input, output_dir)
+
+    # Get all paths that we are supposed to skip before generating the diff and applying updates
+    skip_paths = _get_skip_paths(cruft_state, pyproject_file)
+    # We also get the list of paths that were deleted from the project
+    # directory but were present in the template that the project is linked against
+    # This is to avoid introducing changes that won't apply cleanly to the current project.
+    deleted_paths = _get_deleted_files(output_dir, project_dir)
+    # We now remove skipped and deleted paths from the project
+    _remove_paths(output_dir, skip_paths | deleted_paths)
+
+    return context
+
+
+#####################################
+# Generating clean outputs for diff #
+#####################################
+
+
+def _generate_output(
+    cruft_state: CruftState, project_dir: Path, cookiecutter_input: bool, output_dir: Path
+) -> CookiecutterContext:
+    inner_dir = project_dir / (cruft_state.get("directory") or "")
+
+    new_context = generate_cookiecutter_context(
+        cruft_state["template"],
+        inner_dir,
+        extra_context=cruft_state["context"]["cookiecutter"],
+        no_input=not cookiecutter_input,
+    )
+
+    # This generates the cookiecutter template.
+    # Unfortunately, cookiecutter doesn't let us output the template in an
+    # arbitrary directory. It insists on creating the initial project directory.
+    # Therefore we have to move the directory content to the expected output_dir.
+    # See https://github.com/cookiecutter/cookiecutter/pull/907
+    with TemporaryDirectory() as tmpdir_:
+        tmpdir = Path(tmpdir_)
+
+        # Kindly ask cookiecutter to generate the template
+        template_dir = generate_files(
+            repo_dir=inner_dir, context=new_context, overwrite_if_exists=True, output_dir=tmpdir
+        )
+        template_dir = Path(template_dir)
+
+        # Move the template content to the output directory
```

Comment:
We should not do this. This breaks the expected behavior for the cruft create command. Again similar to my other requests, let's please avoid any changes to other parts of the code expect the refactoring that's needed for diff to work.

Reply:
That's actually useful for contributors like me to be able to reason on the code. I'm ok with making self-contained PRs and everything, but I prefer refactoring an implementation when the existing one would make me write convoluted code. And the current implementation of this on master would have made me write convoluted code, that's why I changed it.

Rather than shaming and looking down on contributors (friendly reminder that I also do this in my spare time, a sunday night, because I like your project), you could try to 1) understand why I did it 2) be precise on what exactly it does break and 3) focus on how we could not-break it while keeping it understandable to human beings.

Reply:
I'm sorry if you got that impression in anyway from my comments. (I even thanked you for the pull request :)) I'm not trying to look down or shame anyone here. I'm not sure what made think that way. I have tried to be cordial and I am extremely sorry if you got this impression.

As for my review - 

This currently breaks the cruft create command as the output directory is no longer determined by the cookiecutter templated directory name. Although this is a minor refactoring for diff and update, this is a major change for users using cruft to create templates. 

This is the same reason why it wasn't just implemented in the upstream pull request directly and instead implemented behind a strip flag.

Lastly, I think your reminder also applies to me, this is not my full time job and I'm doing my best to review this as quickly as possible in my spare time on a Sunday :) 

I also don't have any expectations from you to get back to me with the changes from my review instantly. If it's late, you can get back to this pull request later when you get a chance.

Again I completely appreciate your time and contributions. Please let me know if you need any clarity on any of the other comments I have made.

Reply:
Edit - I saw the code again, sorry my mistake. Looks like it's different function calls. Please ignore this review comment. It makes sense :)

Reply:
I appreciate your reply and I admit that I may have slightly overreacted. What annoyed me was that you seemed to imply and insisted multiple times in the same message that I changed the code for no reason while I tried to refactor it in a way makes sense overall. I had to deeply refactor the `cookiecutter_template` function anyways because updating and diffing were tightly coupled. So that's very different from the 3-4 lines I changed in tests that indeed had nothing to deal with the content of this PR. (By the way, for the latter, I know some maintainers like and even encourage single-line trivial improvements, but I couldn't know you didn't like it before making the PR :smile: ).

Now that you gave a clear and objective statement on what was the problem with this change, I'm just fine with it and I can look for solutions. IMO your first message lacked strong, verifiable, objective arguments which was later fixed by your reply. That's why I found it unfriendly and unwelcoming.

> Lastly, I think your reminder also applies to me, this is not my full time job and I'm doing my best to review this as quickly as possible in my spare time on a Sunday :)

Yup, totally applies to you (that's why I said “also”) and I am very sorry if I made this PR harder to review for you than it should have. I'll try my best to improve that in my further contributions to this project. Most importantly, I do appreciate and am very grateful for your contribution to the open-source community. Cruft is a great and well-thought project. I really hope it'll get the visibility it deserves.